lib/reversetunnel/track/doc.go

```diff
@@ -1,21 +1,67 @@
+successful; otherwise, the Lease should be released with Release() so that a new
+connection can be spawned in its stead. The Lease should also be released if the
+connection fails, or after it's closed.
 
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
+The Tracker keeps track of how many Leases are in unclaimed state, and of which
+proxies have been claimed. In addition, it receives gossip messages (usually
+from the reverse tunnel clients themselves, through the reverse tunnel "gossip"
+protocol, see `handleDiscovery` in lib/reversetunnel/agent.go) containing
+information about proxies in the cluster - namely, which proxies exist, and
+which group ID and generation, if any, they belong to.
+
+A proxy can report a group ID and a group generation (it's valid for a proxy to
+belong to the "" group or the zero generation); the group ID represents
+independent deployments of proxies, the group generation is a monotonically
+increasing counter scoped to each group.
 
-// Package track provides a simple interface for tracking known proxies by
-// endpoint/name and correctly handling expiration and exclusivity.
-// The provided Tracker type wraps a workpool.Pool, updating per-key
-// counts as new proxies are discovered and/or old proxies are expired.
+The Tracker will grant Leases (to attempt to connect to new proxies) based on
+which proxies have been claimed, how many Leases have been granted but have yet
+to claim a proxy (they're still "inflight") and which proxies are known and
+which group ID and generation they have reported:
+
+  - the set of desired proxies is defined to be the union of all proxies in the
+    biggest known generation of each proxy group; if proxies A, B and C belong
+    to the same group but proxy A is in generation 2 while B and C are in
+    generation 1, only proxy A is in the desired set
+  - if in proxy peering mode, the count of desired proxies is limited by the
+    connection count, such that if the connection count is 40 but only 3 proxies
+    are in the desired set, the effective count is going to be 3 - however, if
+    the connection count is 3 and there's 40 proxies in the desired set, the
+    effective count is also going to be 3
+  - the desired proxy count is compared with the sum of the count of currently
+    claimed proxies in the desired set and the current amount of inflight
+    connections (Lease objects that haven't been released and haven't claimed a
+    proxy); new connections can be spawned if the desired proxy count is
+    strictly greater.
+
+Proxies are removed from the tracked set after a TTL (defaulting to 3 minutes),
+or if they're missing in gossip messages from two different proxies.
```

Comment:
It would, but then again that should still be enough to spawn new connections if necessary.

I'll move the logic for early deletes in a followup PR, it's not necessary for proxy groups to work.